components/raftstore/src/store/fsm/peer.rs

```diff
@@ -1915,6 +1900,101 @@ where
+                                }
+                            }
+                            if finished {
+                                warn!(
+                                    "Unsafe recovery, failed voters have already been demoted";
+                                    "region" => ?self.region(),
+                                );
+                                return false;
+                            }
+                            true
+                        };
+
+                        let owned_failed_voters = failed_voters.to_owned();
+                        let owned_shared_state = shared_state.clone();
+                        if still_need_demote() {
+                            info!(
+                                "Unsafe recovery, demoting failed voters after exiting joint stata.";
+                                "region" => ?self.region(),
+                                "failed_voters" => ?owned_failed_voters
+                            );
+                            self.propose_raft_command_internal(
+                                demote_failed_voters_request(
+                                    self.region(),
+                                    &self.fsm.peer.peer,
+                                    owned_failed_voters,
+                                ),
+                                Callback::None,
+                                DiskFullOpt::AllowedOnAlmostFull,
+                            );
+                            self.fsm.peer.unsafe_recovery_state =
+                                Some(UnsafeRecoveryState::DemoteFailedVoters {
+                                    shared_state: owned_shared_state,
+                                    failed_voters: vec![], // No longer needed since here.
+                                    commit_index: self
+                                        .fsm
+                                        .peer
+                                        .raft_group
+                                        .raft
+                                        .raft_log
+                                        .last_index(),
+                                    demote_after_exit: false,
+                                });
+                        } else {
+                            self.fsm.peer.unsafe_recovery_state = None;
+                        }
+                    } else {
+                        info!("Unsafe recovery, exiting joint state."; "region" => ?self.region());
+                        self.propose_raft_command_internal(
+                            exit_joint_request(self.region(), &self.fsm.peer.peer),
+                            Callback::None,
+                            DiskFullOpt::AllowedOnAlmostFull,
+                        );
+
+                        self.fsm.peer.unsafe_recovery_state = None;
+                    }
+                }
+            }
+            Some(_) | None => {}
```

Comment:
_ => {} and add comment for the reason why don't need to check the destroy state.